### PR TITLE
Convert MPIController into a global EDProducer

### PR DESCRIPTION
**Pull request by LucaPacioselli**
#### PR description:

This PR converts `MPIController` into a `global` module allowing multiple transitions to be sent concurrently to `MPISource`. The `MPISource` module has been adapted accordingly.

The main changes are:

* Introduce per-stream cache in `MPIController`, consisting of a dedicated control channel and a pool of data channels. The current implementation uses 3 data channels per stream.
* Redesign `MPISource` around a dedicated control thread and one receiver thread per controller stream assigned to the follower. These threads are separated from framework threads.
* Decouple MPI communication from framework thread by buffering received events in per (run-lumi) queues. Receiver threads populate these `tbb::concurrent_queue`s, while `setRunAndEventInfo()` consumes them.
* Preserve the ordering of luminosity blocks seen by the `MPISource` using the ordered (run-lumi) queues. A queue is considered complete only after the corresponding `globalEndLuminosityBlock()` transition is received from the `MPIController` and will be released after being drained.

#### PR validation:

The code in this PR has been tested with the existing `MPICore` tests. 

To verify that `MPIController` could handle more runs and luminosity blocks at a time, I ran the [controller_cfg.py](https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/MPICore/test/controller_cfg.py) test with `numberOfConcurrentLuminosityBlocks` and `numberOfConcurrentRuns` greater than one, one event per luminosity block, and a number of streams greater than the number of luminosity blocks in a run.

### CODE REVIEW

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test

**Comment by ghyls** [comment]
Could you add a test with some extreme combinations of `streams`, `numberOfConcurrentLuminosityBlocks`, and `numberOfConcurrentRuns`? Just to make sure that the the changes in this PR do not lead to deadlocks, even in corner cases.

Maybe you could start from e.g. 16 streams, perhaps the same `numberOfConcurrentLuminosityBlocks` and, if possible, only one event per luminosity block. If there is a race condition or a deadlock somewhere, these settings would be likely to catch it.

File: HeterogeneousCore/MPICore/plugins/MPIController.cc [hunk 1]
```diff
@@ -204,28 +231,58 @@ MPIController::~MPIController() {
...
-  for (auto& follower : followers_) {
-    follower.sendConnect();
+  unsigned int nStreams = nStreams_;
+  unsigned int nFollowers = followers_.size();
+  streamChannels_.resize(nStreams);
+  channels_.resize(nStreams);
+
+  // Compute distribution
+  std::vector<unsigned int> streamsPerFollower(nFollowers);
+  unsigned int base = nStreams / nFollowers;
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:265]: What would happen here if `nStreams < nFollowers`?

If we have e.g. 4 streams and 6 followers, then the last two followers would get `streamsPerFollower[f] = 0`. Would this cause any errors?

Perhaps we could have a test to test this combination

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 2]
```diff
@@ -194,16 +254,46 @@ MPISource::MPISource(edm::ParameterSet const& config, edm::InputSourceDescriptio
   EDM_MPI_Empty_t buffer;
   MPI_Recv(&buffer, 1, EDM_MPI_Empty, MPI_ANY_SOURCE, EDM_MPI_Connect, comm_, &status);
   edm::LogInfo("MPI") << "connected from " << status.MPI_SOURCE;
+
+  // Receive the number of streams that each follower process will handle
+  controller_.receiveStreamCount(nControllerStreams_);
+  edm::LogInfo("MPI") << "Source will handle " << nControllerStreams_ << " streams";
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:260]: I think this should be a `edm::LogDebug` message

File: HeterogeneousCore/MPICore/plugins/MPIController.cc [hunk 3]
```diff
@@ -204,28 +231,58 @@ MPIController::~MPIController() {
...
+  unsigned int remainder = nStreams % nFollowers;
+  for (unsigned int f = 0; f < nFollowers; ++f) {
+    streamsPerFollower[f] = base + (f < remainder ? 1 : 0);
+    // signal the connection
+    followers_[f].sendConnect();
+
+    // 1) Tell each follower how many controller streams it gets
+    edm::LogInfo("MPI") << "Follower " << f << " gets " << streamsPerFollower[f] << " streams";
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:273]: I think this should be an `edm::LogDebug` message

File: HeterogeneousCore/MPICore/plugins/MPIController.cc [hunk 4]
```diff
@@ -204,28 +231,58 @@ MPIController::~MPIController() {
...
+    followers_[f].sendStreamCount(streamsPerFollower[f]);
+
+    // 2) Create per-slot channels
+    for (unsigned int slot = 0; slot < nStreams; ++slot) {
+      if (f == slot % nFollowers) {
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:278]: Maybe this could be
```suggestion
    for (unsigned int slot = 0; slot < nStreams; slot += nFollowers)
```
so that you don't need the `if`, but also the loop does less iterations

File: HeterogeneousCore/MPICore/plugins/MPIController.cc [hunk 5]
```diff
@@ -204,28 +231,58 @@ MPIController::~MPIController() {
...
+        streamChannels_[slot] = followers_[f].duplicate(slot);
+        channels_[slot].reserve(3);
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:280]: I see this `3` is hardcoded both here and in the Source. Should it better be a configurable parameter of the `MPIController` (maybe with `3` being the default), and then sent to the source e.g. as `sendStreamCount` is sending the number of streams?

**Review by ghyls — COMMENTED** [review]
Some comments I have so far. More will come later!

File: HeterogeneousCore/MPICore/plugins/MPIController.cc (hunk 1 above)
Comment (LucaPacioselli) [reply HeterogeneousCore/MPICore/plugins/MPIController.cc:279]: It should result in an error. I added an exception for the `nStreams < nFollowers` case, indicating that it is not supported.
 
In general, I'm not sure whether it makes sense to support this case.

File: HeterogeneousCore/MPICore/plugins/MPIController.cc (hunk 4 above)
Comment (LucaPacioselli) [reply HeterogeneousCore/MPICore/plugins/MPIController.cc]: Ok thanks, but I think in this case `slot` should start from `f` instead of `0`. I changed it to `for (unsigned int slot = f; slot < nStreams; slot += nFollowers)`

File: HeterogeneousCore/MPICore/plugins/MPIController.cc (hunk 5 above)
Comment (LucaPacioselli) [reply HeterogeneousCore/MPICore/plugins/MPIController.cc]: Added a configurable parameter to the MPIController module to change the number of data channels. Also added a `send/receiveDataChannelsCount` to send it to the MPISource

File: HeterogeneousCore/MPICore/interface/messages.h [hunk 6]
```diff
@@ -22,6 +22,7 @@ enum EDM_MPI_MessageTag {
   EDM_MPI_EndRun,
   EDM_MPI_BeginLuminosityBlock,
   EDM_MPI_EndLuminosityBlock,
+  EDM_MPI_LuminosityBlockComplete,
```
Comment (fwyzard) [inline HeterogeneousCore/MPICore/interface/messages.h:25]: Can you rename these to match the framework transitions?
```suggestion
  EDM_MPI_StreamEndLuminosityBlock,
  EDM_MPI_GlobalEndLuminosityBlock,
```

File: HeterogeneousCore/MPICore/plugins/MPIController.cc [hunk 7]
```diff
@@ -41,22 +42,48 @@
...
+
+  // Per-stream cache: the control channel to this stream's follower plus its data-channel pool
+  struct StreamFollower {
+    std::unique_ptr<MPIChannel> follower;
+    std::vector<std::unique_ptr<MPIChannel>> dataChannels;
+  };
+}  // namespace
+
+class MPIController
+    : public edm::global::EDProducer<edm::StreamCache<StreamFollower>, edm::LuminosityBlockCache<EmptyLumiCache>> {
```
Comment (fwyzard) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:57]: Would
```suggestion
    : public edm::global::EDProducer<edm::StreamCache<StreamFollower>, edm::LuminosityBlockCache<void>> {
```
work ?

According to [the documentation for `edm::LuminosityBlockCache`](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface#edm_LuminosityBlockCacheT):

> Note that the template argument `void` is an allowed value and can be used in the case where you just want to be informed about the `globalEndLuminosityBlock` and do not need to share any information across the Stream instances.

File: HeterogeneousCore/MPICore/plugins/MPIController.cc [hunk 8]
```diff
@@ -73,8 +100,9 @@ class MPIController : public edm::one::EDProducer<edm::one::WatchRuns, edm::one:
   }
 
   MPI_Comm comm_ = MPI_COMM_NULL;
-  std::vector<MPIChannel> followers_;
-  std::vector<std::vector<std::unique_ptr<MPIChannel>>> channels_;
+  mutable std::vector<MPIChannel> followers_;
+  mutable std::vector<std::vector<std::unique_ptr<MPIChannel>>> channels_;
+  mutable std::vector<std::unique_ptr<MPIChannel>> streamChannels_;
```
Comment (fwyzard) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:105]: Why do these need to be `mutable`?

Can you document it here?

File: HeterogeneousCore/MPICore/plugins/MPIController.cc [hunk 9]
```diff
@@ -204,28 +239,64 @@ MPIController::~MPIController() {
...
+      }
     }
   }
 
-  // Disconnect the per-follower communicators.
   for (auto& follower : followers_) {
     follower.reset();
   }
 
-  // Close the intercommunicator.
```
Comment (fwyzard) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:219]: You didn't like these comments ?

File: HeterogeneousCore/MPICore/plugins/MPIController.cc (hunk 1 above)
Comment (fwyzard) [reply HeterogeneousCore/MPICore/plugins/MPIController.cc:279]: I guess it _might_ make sense if the association stream-to-follower is not hard coded.

If it is hardcoded the remaining followers would never receive any data, so it would not be very useful to have them.

Maybe they could be repurposed as "back up" servers ?

Anyway, I guess it's OK to require the number of followers to be equal or less than the number of streams.

Can you please update the [`README.md`](https://github.com/cms-sw/cmssw/tree/master/HeterogeneousCore/MPICore/README.md) with this information ?

Comment (fwyzard) [reply HeterogeneousCore/MPICore/plugins/MPIController.cc:279]: Actually after reading the next part I have to ask... could you document in the `README.md` the whole logic :-D ?

File: HeterogeneousCore/MPICore/plugins/MPIController.cc [hunk 10]
```diff
@@ -370,31 +463,24 @@ void MPIController::produce(edm::Event& event, edm::EventSetup const& setup) {
...
-      slot = i;
-      found = true;
-      break;
+  while (!found) {
+    for (unsigned int i = 0; i < channels.size(); ++i) {
+      if (channels[i]->ready()) {
+        slot = i;
+        found = true;
+        break;
+      }
```
Comment (fwyzard) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:477]: The current code
```c++
  bool found = false;
  while (!found) {
    for (unsigned int i = 0; i < channels.size(); ++i) {
      if (channels[i]->ready()) {
        slot = i;
        found = true;
        break;
      }
    }
  }
```
does not actually use the value of `found` after the loop, so it is equivalent to
```c++
  for (unsigned int i = 0; i < channels.size(); ++i) {
    if (channels[i]->ready()) {
      slot = i;
      break;
    }
  }
```

However, what happens if none of the channels are ready ?

Comment (fwyzard) [reply HeterogeneousCore/MPICore/plugins/MPIController.cc:477]: For example, won't `*channels[slot]` at line 401/487 access some invalid object?

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 11]
```diff
@@ -3,6 +3,15 @@
 #include <stdexcept>
 #include <string>
 #include <vector>
+#include <condition_variable>
+
+#include <atomic>
+#include <map>
+#include <mutex>
+#include <thread>
```
Comment (fwyzard) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:11]: Could you keep the includes sorted ?
```suggestion
#include <atomic>
#include <condition_variable>
#include <map>
#include <memory>
#include <mutex>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>
```

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 12]
```diff
@@ -305,12 +468,11 @@ bool MPISource::setRunAndEventInfo(edm::EventID& event,
...
       // BeginLuminosityBlock message
       case EDM_MPI_BeginLuminosityBlock: {
-        // receive the LuminosityBlockAuxiliary
         EDM_MPI_LuminosityBlockAuxiliary_t buffer;
         MPI_Mrecv(&buffer, 1, EDM_MPI_LuminosityBlockAuxiliary, &message, &status);
         // TODO this is currently not used
-        edm::LuminosityBlockAuxiliary luminosityBlockAuxiliary;
-        edmFromBuffer(buffer, luminosityBlockAuxiliary);
+        edm::LuminosityBlockAuxiliary aux;
+        edmFromBuffer(buffer, aux);
```
Comment (fwyzard) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:475]: This part could stay unchanged ?

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 13]
```diff
@@ -349,43 +493,148 @@ bool MPISource::setRunAndEventInfo(edm::EventID& event,
       case EDM_MPI_ProcessEvent: {
         // receive the EventAuxiliary
         edm::EventAuxiliary aux;
-        unsigned int slot;
-        status = controller_.receiveEvent(aux, slot, message);
+        unsigned int ctrlSlot;
```
Comment (fwyzard) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:496]: Can you keep `slot` ?
Or would it conflict with anything else ?

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 14]
```diff
@@ -237,52 +355,97 @@ bool MPISource::setRunAndEventInfo(edm::EventID& event,
...
-        history_.clear();
-        controller_.receiveProduct(0, history_);
-        history_.initializeTransients();
+        // Protect shared history_ access now that MPIController is a global module and receive the ProcessHistory
+        {
+          std::lock_guard<std::mutex> lock(historyMutex_);
+          history_.clear();
+          worker.ctrlStream->receiveProduct(0, history_);
+          history_.initializeTransients();
+        }
```
Comment (fwyzard) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:448]: I'm not sure what to do about this part.

Initially I added the support for the process history assuming we would need it to later implement support for the `edm::Ref`.

But we don't really know how to support the `edm::Ref`, or even if it will be possible in the short term, and so support for the history was unused.

Then, looking at the other uses of `registerProcessHistory()`, I have the impression that they are always called in the _constructor_ of a `Source`, not while processing new Runs.

I guess one option could be to keep the minimal support: uncomment the call to `processHistoryRegistryForUpdate().registerProcessHistory(history_)`, change the `LogInfo` to `LogDebug`, and move them inside the mutex block.

The other option is to just remove sending and receiving the `ProcessHistory`. I would still keep the MPI messages and the serialisation support, and comment out the code in the `MPIController` and `MPISource` with a message about adding support later. 

All in all, I'm leaning towards the latter.

@ghyls @LucaPacioselli @makortel opinions?

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 15]
```diff
@@ -349,43 +493,148 @@ bool MPISource::setRunAndEventInfo(edm::EventID& event,
...
+
+    if (status.MPI_TAG != EDM_MPI_SendModulesInfo) {
+      throw cms::Exception("InvalidValue") << "The MPISource has received an unexpected message with tag "
+                                           << status.MPI_TAG << " while expecting EDM_MPI_SendModulesInfo";
+    }
+
+    int size;
+    MPI_Get_count(&status, MPI_BYTE, &size);
+    std::vector<char> buffer;
+    buffer.resize(size);
```
Comment (fwyzard) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:565]:
```suggestion
    std::vector<char> buffer(size);
```
?

File: HeterogeneousCore/MPICore/plugins/MPIController.cc [hunk 16]
```diff
@@ -83,19 +111,28 @@ class MPIController : public edm::one::EDProducer<edm::one::WatchRuns, edm::one:
...
+  unsigned int nStreams_{0};
+  unsigned int nDataChannels_{0};
 };
 
 MPIController::MPIController(edm::ParameterSet const& config)
     : token_(produces<MPIToken>()),
       mode_(parseMode(config.getUntrackedParameter<std::string>("mode"))),
-      module_label_(config.getParameter<std::string>("@module_label"))  //
+      module_label_(config.getParameter<std::string>("@module_label")),
+      nDataChannels_(config.getUntrackedParameter<unsigned int>("numberDataChannels", 3))  //
```
Comment (fwyzard) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:122]: Rather than making this configurable, can we start with a single channel and increase the number at runtime, if needed ?

File: HeterogeneousCore/MPICore/plugins/MPISource.cc (hunk 14 above)
Comment (makortel) [reply HeterogeneousCore/MPICore/plugins/MPISource.cc:448]: > Then, looking at the other uses of `registerProcessHistory()`, I have the impression that they are always called in the _constructor_ of a `Source`, not while processing new Runs.

`PoolSource` seems to call it when opening new input files (because new files could have different histories).

I'd have to dig deeper and digest more to be able to comment more.

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 17]
```diff
@@ -237,52 +355,97 @@ bool MPISource::setRunAndEventInfo(edm::EventID& event,
...
+void MPISource::markLumiComplete(unsigned int run, unsigned int lumi) {
+  std::lock_guard<std::mutex> lock(mapMutex_);
+  auto it = lumiQueues_.find({run, lumi});
+  if (it != lumiQueues_.end()) {
+    it->second->complete = true;
+    mainCV_.notify_one();
+  } else {
+    // create it to avoid missing the completion
+    auto q = std::make_shared<LumiQueue>();
+    q->complete = true;
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:423]: Don't we need a `mainCV_.notify_one();` here as well?
(or to move the one above after the `if`)

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 18]
```diff
@@ -237,52 +355,97 @@ bool MPISource::setRunAndEventInfo(edm::EventID& event,
...
-        EDM_MPI_Empty_t buffer;
-        MPI_Mrecv(&buffer, 1, EDM_MPI_Empty, &message, &status);
-
-        // receive the next message
-        break;
+      // unexpected message
+      default: {
+        throw cms::Exception("InvalidValue")
+            << "The MPISource has received an unknown message with tag " << status.MPI_TAG;
+        return;
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:380]: Do we need to `return` after `throw`ing? 
I see it in other places as well

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 19]
```diff
@@ -194,16 +255,48 @@ MPISource::MPISource(edm::ParameterSet const& config, edm::InputSourceDescriptio
...
+      if (worker->ctrlStream)
+        worker->ctrlStream->reset();
+      for (auto& c : worker->dataChannels) {
+        if (c)
+          c->reset();
+      }
+    }
+  }
+
+  // Stop control thread
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:295]: The worker threads above are produced by `controlThread_` in `&MPISource::controlThreadLoop`.

I don't think it is be possible that these workers are destroyed while the `contrtolThread_` is creating them. But just in case, maybe it's better to stop the `controlThread_` first and then the workers

File: HeterogeneousCore/MPICore/plugins/MPIController.cc
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:2]: Could you `#include <memory>` here? 
For the `std::unique_ptr` used below

File: HeterogeneousCore/MPICore/plugins/MPIController.cc [hunk 20]
```diff
@@ -204,28 +239,64 @@ MPIController::~MPIController() {
...
+  if (nStreams < nFollowers) {
+    throw edm::Exception(edm::errors::Configuration)
+        << "ERROR: Number of Followers is greater than number of Controller Streams is not supported.";
+  }
+
+  streamChannels_.resize(nStreams);
+  channels_.resize(nStreams);
+
+  // Compute distribution
+  std::vector<unsigned int> streamsPerFollower(nFollowers);
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:278]: Could this just be a `unsigned int` inside the loop?
Since I see it is not used anywhere outside the loop below

File: HeterogeneousCore/MPICore/plugins/MPIController.cc [hunk 21]
```diff
@@ -204,28 +239,64 @@ MPIController::~MPIController() {
...
+  unsigned int base = nStreams / nFollowers;
+  unsigned int remainder = nStreams % nFollowers;
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:280]: (Maybe pedantic)
(`base` and `remainder` can be `const`)

File: HeterogeneousCore/MPICore/plugins/MPIController.cc [hunk 22]
```diff
@@ -254,7 +325,29 @@ void MPIController::endJob() {
...
+  // Reclaim the per-stream channels before the framework destroys the cache.
+  // MPISource only disconnects its matching per-stream channels inside its
+  // destructor, after it has received the global Disconnect message. Resetting
+  // here would block waiting for a follower action that's gated on endJob()
+  // being called, which won't happen until every endStream() returns.
+  streamChannels_[sid.value()] = std::move(cache.follower);
+  channels_[sid.value()] = std::move(cache.dataChannels);
+}
+
+void MPIController::streamBeginRun(edm::StreamID sid, edm::Run const& run, edm::EventSetup const& setup) const {
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPIController.cc:350]:
```suggestion
void MPIController::streamBeginRun(edm::StreamID sid, edm::Run const& run, edm::EventSetup const&) const {
```

Since `setup` is not used

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 23]
```diff
@@ -237,52 +355,97 @@ bool MPISource::setRunAndEventInfo(edm::EventID& event,
...
-        EDM_MPI_Empty_t buffer;
-        MPI_Mrecv(&buffer, 1, EDM_MPI_Empty, &message, &status);
+void MPISource::startThread() {
+  workers_.push_back(std::make_unique<StreamWorker>());
+  StreamWorker& worker = *workers_.back();
 
-        // receive the next message
-        break;
-      }
+  worker.thread_id = nextThreadID_;
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:390]: Do we need this `thread_id`? Does not seem to be used

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 24]
```diff
@@ -54,6 +63,37 @@ class MPISource : public edm::ProducerSourceBase {
...
+
+  struct LumiQueue {
+    tbb::concurrent_queue<EventItem> events;
+    // set to true when LumiComplete received
+    std::atomic<bool> complete{false};
+  };
+
+  using RunLumi = std::pair<unsigned int, unsigned int>;
+
+  unsigned int nextThreadID_ = 0;
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:80]: I would personally move this down with the other data members

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 25]
```diff
@@ -68,13 +108,34 @@ class MPISource : public edm::ProducerSourceBase {
...
   Mode mode_;
+
+  std::mutex historyMutex_;
   edm::ProcessHistory history_;
 
-  // temporary value used to pass information from setRunAndEventInfo() to produce()
   MPIChannel* channel_ = nullptr;
+
+  unsigned int nControllerStreams_{0};
+  unsigned int nDataChannels_{0};
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:121]: ( As a general rule, it's a good practice to roughly sort data members in decreasing size. It makes classes more compact in memory. Not that it matters that much here. )

File: HeterogeneousCore/MPICore/plugins/MPISource.cc [hunk 26]
```diff
@@ -194,16 +255,48 @@ MPISource::MPISource(edm::ParameterSet const& config, edm::InputSourceDescriptio
...
+
+  // Receive the number of streams that each follower process will handle and the number of
+  // data channels available per controller stream
+  controller_.receiveStreamCount(nControllerStreams_);
+  LogDebug("MPI") << "Source will handle " << nControllerStreams_ << " streams";
+  controller_.receiveDataChannelCount(nDataChannels_);
+
+  // Each controller stream gets an independent control channel plus one data
+  // channel per controller slot (WIP: fixed 3). The duplicated communicators allow different
+  // receiver threads to operate without sharing MPIChannel instances.
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/plugins/MPISource.cc:267]: Remember to update the comment once we decide on the strategy here :)

File: HeterogeneousCore/MPICore/src/MPIChannel.cc [hunk 27]
```diff
@@ -227,6 +227,12 @@ void MPIChannel::edmToBuffer_(EDM_MPI_EventAuxiliary_t& buffer, edm::EventAuxili
   buffer.slotId = slot;
 }
 
+// blocking probe for the next message on this channel's communicator
+MPI_Status MPIChannel::probeAny(MPI_Message& message, MPI_Status& status) const {
+  MPI_Mprobe(MPI_ANY_SOURCE, MPI_ANY_TAG, comm_, &message, &status);
+  return status;
+}
```
Comment (ghyls) [inline HeterogeneousCore/MPICore/src/MPIChannel.cc:234]: We are giving back `status` twice here, as the return value and via `MPI_Status& status`.
Maybe just keep the return value, and drop the `MPI_Status& status` from the function signature?

**Review by ghyls — COMMENTED** [review]
Some additional comments from my side!
